### PR TITLE
Add a 3D Box regression loss for CenterPillar

### PR DESCRIPTION
**Pull request by ianstenbit**
This loss is used for box regression in the CenterPillar model, and is the last component missing in KerasCV to have an end-to-end trainable 3D OD model

### CODE REVIEW

File: keras_cv/losses/box_3d_regression_loss.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+import math
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_cv.bounding_box_3d import CENTER_XYZ_DXDYDZ_PHI
+
+
+def l1(y_true, y_pred, sigma=9.0):
```
Comment (LukeWood) [inline keras_cv/losses/box_3d_regression_loss.py:23]: can we use a keras L1 loss component to do this, just for standardization? 

Consider this a nit - not a blocker.

File: keras_cv/losses/box_3d_regression_loss.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    loss = tf.where(
+        absolute_difference < 1.0 / sigma,
+        0.5 * sigma * absolute_difference**2,
+        absolute_difference - 0.5 / sigma,
+    )
+
+    return tf.reduce_sum(loss, axis=-1)
+
+
+class Box3DRegressionLoss(keras.losses.Loss):
```
Comment (LukeWood) [inline keras_cv/losses/box_3d_regression_loss.py:36]: do you like the name Box3D?  I feel like theres a bit of an inversion with ordering here (3D box is my expected ordering), but I understand python probably prevents the name 3DBox.  

Wonder if there's anything else we can call these components.

No strong opinion, and honestly no better name in mind, just thought I'd mention it in case you have a better anme.

File: keras_cv/losses/box_3d_regression_loss.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    """Implements a bin-based box regression loss for 3D bounding boxes.
+
+    Reference: https://arxiv.org/abs/1812.04244
+
+    Box3DRegressionLoss uses L1 loss on the individual components of boxes, with
```
Comment (LukeWood) [inline keras_cv/losses/box_3d_regression_loss.py:41]: are these delta encoded?  if so, good to clarify this.

File: keras_cv/losses/box_3d_regression_loss.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                * heading_bin_one_hot,
+                axis=-1,
+                keepdims=True,
+            ),
+            heading_res_norm_label[..., tf.newaxis],
+        )
+
+        return loss_heading_bin + loss_heading_res
+
+    def regression_loss(self, y_true, y_pred):
```
Comment (LukeWood) [inline keras_cv/losses/box_3d_regression_loss.py:98]: I feel like this class is doing 2~ things at once:

- unpackaging formats
- computing loss on these formats.

I feel something analogous to this is in the RetinaNet's compile method.  Instead of having a generic `BoxRegressionLoss`, we take 2 losses to compile(), `box_loss`, `classification_loss`.  Is the reason that this is packaged together due to inter-dependencies?

Perhaps it can be refactored to something like:

```
Box3DLoss(heading_loss='...', box_loss='...',)
```

Or we can attach some of the logic to the CenterPillar model instead.

Thoughts?  Am I missing some context as to why this wouldn't work, or why it is optimal to keep it all in one place?

File: keras_cv/losses/box_3d_regression_loss.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    """Computes element-wise l1 loss."""
+
+    absolute_difference = tf.abs(y_pred - y_true)
+    loss = tf.where(
+        absolute_difference < 1.0 / sigma,
+        0.5 * sigma * absolute_difference**2,
+        absolute_difference - 0.5 / sigma,
+    )
+
+    return tf.reduce_sum(loss, axis=-1)
```
Comment (LukeWood) [inline keras_cv/losses/box_3d_regression_loss.py:33]: can you just remove this and then remove the `tf.new_axis` used below (along with `reduce_sum(keep_axis=True)`?

**Review by LukeWood — CHANGES_REQUESTED** [review]
Did a first pass, my first instinct is that this is handling 2-3 distinct responsibilities in one component.  Perhaps we can break it down into more modular building blocks.

(I mentioned this in a comment too, just to echo I've copy pasted that comment below):

For clarity/to echo the comment:

I feel like this class is doing 2~ things at once:

unpackaging formats
computing loss on these formats.
I feel something analogous to this is in the RetinaNet's compile method. Instead of having a generic BoxRegressionLoss, we take 2 losses to compile(), box_loss, classification_loss. Is the reason that this is packaged together due to inter-dependencies?

Perhaps it can be refactored to something like:

Box3DLoss(heading_loss='...', box_loss='...',)
Or we can attach some of the logic to the CenterPillar model instead.

Thoughts? Am I missing some context as to why this wouldn't work, or why it is optimal to keep it all in one place?

File: keras_cv/losses/box_3d_regression_loss.py (hunk 2 above)
Comment (ianstenbit) [reply keras_cv/losses/box_3d_regression_loss.py:36]: Yeah, I wanted to call it `3DBoxRegressionLoss`, but Python doesn't like starting with the `3`

File: keras_cv/losses/box_3d_regression_loss.py (hunk 4 above)
Comment (ianstenbit) [reply keras_cv/losses/box_3d_regression_loss.py:98]: This is a bit of a complex issue -- I agree we should think about it.

CenterNet will already take two losses:
- Heatmap loss
- Box regression loss

IMO it makes sense for this to be one loss, because fundamentally the whole thing represents a regression loss for 3D boxes. The reason that the headings are handled separately is simply that the headings are predicted using bins instead of a single regression target, which makes the code more complex.

Happy to chat about this more. There's also an additional concept that is sometimes added to the box loss (which we'll want to support) which is a 3D IoU component. In that case, it may make sense to include this as a separate loss function.

File: keras_cv/losses/box_3d_regression_loss.py (hunk 1 above)
Comment (ianstenbit) [reply keras_cv/losses/box_3d_regression_loss.py:23]: Keras doesn't offer an L1 loss component, only a regularizer (which is a different function)

File: keras_cv/losses/box_3d_regression_loss.py (hunk 3 above)
Comment (ianstenbit) [reply keras_cv/losses/box_3d_regression_loss.py:41]: Updated to clarify the encoding a bit

File: keras_cv/losses/box_3d_regression_loss.py (hunk 5 above)
Comment (ianstenbit) [reply keras_cv/losses/box_3d_regression_loss.py:33]: The `l1` function is called from a few different spots, and this would only work for the heading residual, not the location and size components

**Comment by ianstenbit** [comment]
With respect to doing 2 things, I'm not sure it makes any sense to split this up.

The only purpose of decoding the predictions (which we only do a little bit of here) is to correctly compute loss, and it's not relevant to other parts of the model. This loss is all conceptually 1 thing, which is regressing against ground truth boxes, whereas e.g. Heatmap loss in CenterNet is a separate concept (and is thus handled separately in the model)

**Comment by ianstenbit** [comment]
@LukeWood -- renamed this to make it more over that it's really only useful for CenterPillar, and I also updated the docstring to reflect this.

File: keras_cv/losses/box_3d_regression_loss.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/losses/box_3d_regression_loss.py]: (As discussed in 1:1 let’s call it CenterPillarLoss)

**Comment by LukeWood** [comment]
> @LukeWood -- renamed this to make it more over that it's really only useful for CenterPillar, and I also updated the docstring to reflect this.

Cool, will do another pass

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
Thanks @ianstenbit - with the rename and all now it is clear how it is used and it all looks logical to me.
